### PR TITLE
Rework ZCash support

### LINKED ISSUE #13: Update Zcash address encoding / decoding
Same as Bitcoin lockscript

### PR DESCRIPTION
**Pull request by catenocrypt**
Fixes #13 .
Factored out Base58Address, and used in other address schemes.

### CODE REVIEW

File: zcash.go [hunk 1]
```diff
@@ -19,40 +24,10 @@ func init() {
...
-	decoded, err := Base58ChecksumDecode(input, Base58DefaultAlphabet)
-	if err != nil {
-		return nil, err
-	}
-	err = checkIsValid(decoded)
-	if err != nil {
-		return nil, err
-	}
-	return decoded, nil
+	return Base58AddressDecodeToBytesPrefix(input, ZCashAddressLength, zcashPrefixes)
```
Comment (hewigovens) [inline zcash.go:27]: we should decode to script like bitcoin: https://github.com/trustwallet/ens-coincodec/blob/master/bitcoin_test.go#L15

**Review by hewigovens — COMMENTED** [review]
test coverage dropped 1.55%

**Comment by catenocrypt** [comment]
> test coverage dropped 1.55%

I saw that but it is hard to believe, as I replaced much code with common code.

File: zcash.go (hunk 1 above)
Comment (catenocrypt) [reply zcash.go]: Reworked as you suggested.  In fact ZCash is now a bitcoin fork, just like the others.

**Comment by catenocrypt** [comment]
To check coverage

**Comment by catenocrypt** [comment]
Coverage improved 
+ Coverage   94.53%   95.91%   +1.37%

File: bitcoin.go [hunk 2]
```diff
@@ -70,43 +69,56 @@ func bitcoinDecodeToBytes(input string, config *CoinConfig) ([]byte, error) {
...
+	data, err := Base58AddressDecodeToBytesPrefix(input, btcKeyHashLenght+len(config.P2SHPPrefix), [][]byte{config.P2PKHPrefix, config.P2SHPPrefix})
 	if err != nil {
 		if len(config.HRP) <= 0 {
 			return nil, err
 		}
 		// try bech32
-		decodedHrp, bytes, err := bech32.Decode(input)
-		if err != nil {
-			return nil, errors.Wrapf(err, "decoding bech32 failed")
+		decodedHrp, data, err2 := bech32.Decode(input)
```
Comment (hewigovens) [inline bitcoin.go:78]: it's ok to return only last error
```suggestion
		decodedHrp, data, err := bech32.Decode(input)
```

File: bitcoin.go [hunk 3]
```diff
@@ -70,43 +69,56 @@ func bitcoinDecodeToBytes(input string, config *CoinConfig) ([]byte, error) {
...
-	// check version byte
-	if version == config.P2PKHPrefix {
-		return buildP2PKHScript(bytes), nil
-	} else if version == config.P2SHPPrefix {
-		return buildP2SHScript(bytes), nil
+	// config.P2SHPPrefix)
+	return buildP2SHScript(data), nil
+}
+
+func replacePrefix(input []byte, oldPrefix []byte, newPrefix []byte) []byte {
```
Comment (hewigovens) [inline bitcoin.go:100]: what does this method do?

File: bitcoin_forks_test.go [hunk 4]
```diff
@@ -8,7 +8,7 @@ import (
 )
```
Comment (hewigovens) [inline bitcoin_forks_test.go:9]: no Zcash tests in `bitcoin_forks_test`

**Review by hewigovens — APPROVED** [review]
looks good, left some comments

File: bitcoin_forks_test.go (hunk 4 above)
Comment (catenocrypt) [reply bitcoin_forks_test.go:9]: Oops, they were lost, added back.  Good catch.

File: bitcoin.go (hunk 3 above)
Comment (catenocrypt) [reply bitcoin.go:100]: Removes a []byte prefix from input data if present, and adds another prefix instead.  Used to remove the opcode prefix like OP_DUP,OP_HASH160,20 and replace with P2P prefix like []byte{0x05}.  P2P prefix is one byte for most BTC forks, except ZCash, where it is 2.

File: bitcoin.go (hunk 2 above)
Comment (catenocrypt) [reply bitcoin.go:78]: In this 'try both decoding version and see if one succeed', it is useful to have both errors.  Particularly, in the wrong prefix case we get the bech32 error, which is not very useful.
But let's keep it simple.

**Comment by catenocrypt** [comment]
Coverage   94.53%   96.12%   +1.59%

**Pull request merged**